src/Microsoft.Restier.Core/Api.cs

```diff
@@ -78,16 +78,40 @@ public static class Api // TODO GitHubIssue#25,#26 : transactions, exception fil
         {
             Ensure.NotNull(context, "context");
             var config = context.Configuration;
-            if (config.Model == null)
+            if (config.Model != null)
             {
-                var builder = context.GetApiService<IModelBuilder>();
-                if (builder != null)
-                {
-                    config.Model = await builder.GetModelAsync(new InvocationContext(context), cancellationToken);
-                }
+                return config.Model;
+            }
+
+            var builder = context.GetApiService<IModelBuilder>();
+            if (builder == null)
+            {
+                return null;
+            }
+
+            Task<IEdmModel> running;
+            var source = config.CompeteModelGeneration(out running);
+            if (source == null)
+            {
+                return await running;
             }
 
-            return config.Model;
+            try
+            {
+                var model = await builder.GetModelAsync(new InvocationContext(context), cancellationToken);
+                source.SetResult(model);
+                return model;
+            }
+            catch (AggregateException e)
```

Comment:
If we catch AggregatedException as normal Exception, and SetException to the task, the task will wrap it again and so ends with an AggreatedException, which has an inner AggregatedException, which has the actual exceptions as its InnerExceptions.